### PR TITLE
Reduce memory overhead generated by kmem_cache_create

### PR DESCRIPTION
**Pull request by Adam-pi3**
This commit brings a few important changes:
 - LKRG has used to leverage SLAB_HWCACHE_ALIGN but memory overhead
   may be too significant for LKRG's use cases
 - Since the kernel 4.5+ we can use SLAB_ACCOUNT to make sure that
   LKRG's caches are standalone
 - Modify the size of pCFI stack buffer cache to be smaller and
   decoupled from the PAGE_SIZE (there is no reason for that)

Additionally, this commit should help addressing #131

### CODE REVIEW

**Comment by Adam-pi3** [comment]
Old cache sizes:
```
p_ed_pcfi_cache - 2122
p_ed_pids - 448
p_ed_wq_valid_cache - 64
p_offload_cache - 64
```
New cache sizes:
```
p_ed_pcfi_cache - 520
p_ed_pids - 408
p_ed_wq_valid_cache - 40
p_offload_cache - 40
```

File: src/modules/exploit_detection/p_exploit_detection.h [hunk 1]
```diff
@@ -232,9 +232,9 @@ struct p_seccomp {
 };
 
 #ifdef P_LKRG_TASK_OFF_DEBUG
- #define P_PCFI_STACK_BUF     (PAGE_SIZE >> 2)
+ #define P_PCFI_STACK_BUF     0x150
 #else
- #define P_PCFI_STACK_BUF     (PAGE_SIZE >> 1)
+ #define P_PCFI_STACK_BUF     0x200
 #endif
```
Comment (solardiz) [inline src/modules/exploit_detection/p_exploit_detection.h:238]: What exactly happens if these turn out to be insufficient? Perhaps you also tested with smaller values to see that?

File: src/p_lkrg_main.h [hunk 2]
```diff
@@ -103,6 +103,20 @@
...
+ *  - LKRG has used to leverage SLAB_HWCACHE_ALIGN but memory overhead
+ *    may be too significant for LKRG's use cases
+ *  - Since the kernel 4.5+ we can use SLAB_ACCOUNT to make sure
+ *    that LKRG's caches are standalone
+ */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,5,0)
+ #define P_LKRG_CACHE_FLAGS 0
+#else
+ #define P_LKRG_CACHE_FLAGS SLAB_ACCOUNT
+#endif
```
Comment (solardiz) [inline src/p_lkrg_main.h:117]: I wonder if we can/should use `#ifdef SLAB_ACCOUNT` in place of the kernel version check.

**Review by solardiz — APPROVED** [review]

**Comment by solardiz** [comment]
> use SLAB_ACCOUNT to make sure that LKRG's caches are standalone

Do we need them to be standalone, and why?

File: src/modules/exploit_detection/p_exploit_detection.h (hunk 1 above)
Comment (Adam-pi3) [reply src/modules/exploit_detection/p_exploit_detection.h:238]: We will not validate all the addresses on the stack. On x86-64 we are validating up to 64 entries in stack depth at the time of validation. If stack depth are deeper we would only validate up to 64 entries (or 128 entries on 32bits x86).

Comment (Adam-pi3) [reply src/modules/exploit_detection/p_exploit_detection.h:238]: Now I see that for CONFIG_UNWINDER_ORC we don't have this check (but for non CONFIG_UNWINDER_ORC). I will add this corner case validation for CONFIG_UNWINDER_ORC as well

File: src/p_lkrg_main.h (hunk 2 above)
Comment (Adam-pi3) [reply src/p_lkrg_main.h:117]: I would prefer to keep it as is since we might want to use other flags in the future.

**Comment by Adam-pi3** [comment]
> > use SLAB_ACCOUNT to make sure that LKRG's caches are standalone
> 
> Do we need them to be standalone, and why?

LKRG's objects can't be used in exploitation and in case we have bug, it is difficult to exploit it.

File: src/modules/exploit_detection/p_exploit_detection.h (hunk 1 above)
Comment (solardiz) [reply src/modules/exploit_detection/p_exploit_detection.h:238]: > I will add this corner case validation for CONFIG_UNWINDER_ORC as well

Do you intend to add one more commit to this PR, or create a separate PR?

Also, ideally we'd test triggering the condition in both kinds of builds (by temporarily using deliberately way too low values of `P_PCFI_STACK_BUF`).

Comment (Adam-pi3) [reply src/modules/exploit_detection/p_exploit_detection.h:238]: I've just pushed extra commit to this PR. I've tested this condition for compilation with CONFIG_UNWINDER_ORC and without it. In both cases it works fine - up to N frames from the top are validated (N is related to `P_PCFI_STACK_BUF`).

File: src/p_lkrg_main.h (hunk 2 above)
Comment (solardiz) [reply src/p_lkrg_main.h:117]: OK, but even for other flags I think `#ifdef FLAG_NAME` will work better (as long as the kernel defines all flags as macros) - it would enable usage of the flags precisely where available, including in distro backports, with no extra effort on our part.

File: src/modules/exploit_detection/p_exploit_detection.h (hunk 1 above)
Comment (solardiz) [reply src/modules/exploit_detection/p_exploit_detection.h:238]: The CONFIG_UNWINDER_ORC case fix looks good to me.

File: src/p_lkrg_main.h (hunk 2 above)
Comment (solardiz) [reply src/p_lkrg_main.h:117]: Anyway, I think I'll merge this as-is now, and we can switch to checking by flag names in a later code cleanup.
